### PR TITLE
[feat] allow SDXL pipeline to run with fused QKV projections

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

Adds an option for running the SDXL pipeline with QKV projections fused. For self-attention, all the projection matrices are horizontally fused. For cross-attention, key and value projection matrices are fused. 

Some more comments are inline. 

A lot of thanks to @cpuhrsch for helping.

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py [hunk 1]
```diff
@@ -718,6 +718,67 @@ def disable_freeu(self):
...
+        """
+        from ...models.attention_processor import FusedAttnProcessor2_0
+
+        self.fusing_unet = False
+        self.fusing_vae = False
+
+        if unet:
+            self.fusing_unet = True
+            self.unet.enable_fused_qkv_projections(device=self.device, dtype=self.dtype)
+            self.unet.set_attn_processor(FusedAttnProcessor2_0())
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:744]: I don't think it makes sense to add variants of this attention processor (xformers, PT 1.x, etc.) for now as that way we can push for pure native PT 2.0. Also, it tends to show benefit when you use `torch.compile()`. So, either way, keeping it to PT 2.0 only, for now, makes more sense to me.

File: src/diffusers/schedulers/scheduling_euler_discrete.py [hunk 2]
```diff
@@ -289,6 +290,8 @@ def set_timesteps(self, num_inference_steps: int, device: Union[str, torch.devic
             self.timesteps = torch.from_numpy(timesteps.astype(np.float32)).to(device=device)
 
         self.sigmas = torch.cat([sigmas, torch.zeros(1, device=sigmas.device)])
+        if "cuda" in str(sigmas.device):
+            self.sigmas = self.sigmas.tolist()
```
[the remark this answers is not in the document]
Comment (charchit7) [reply src/diffusers/schedulers/scheduling_euler_discrete.py:294]: ohh, that's interesting.

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py [hunk 3]
```diff
@@ -670,8 +670,9 @@ def _get_add_time_ids(
         add_time_ids = torch.tensor([add_time_ids], dtype=dtype)
         return add_time_ids
 
-    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion_upscale.StableDiffusionUpscalePipeline.upcast_vae
     def upcast_vae(self):
+        from ...models.attention_processor import FusedAttnProcessor2_0
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:674]: Why not import at the very top?

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py [hunk 4]
```diff
@@ -718,6 +720,67 @@ def disable_freeu(self):
...
+            vae (`bool`, defaults to `True`): To apply fusion on the VAE.
+        """
+        from ...models.attention_processor import FusedAttnProcessor2_0
+
+        self.fusing_unet = False
+        self.fusing_vae = False
+
+        if unet:
+            self.fusing_unet = True
+            self.unet.enable_fused_qkv_projections(device=self.device, dtype=self.dtype)
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:745]:
```suggestion
            self.unet.fuse_qkv_projections(device=self.device, dtype=self.dtype)
```
(nit) think "fuse_..." is a bit more readable than "enable_fused_..."

Why is device and dtype passed?

File: src/diffusers/models/attention_processor.py [hunk 5]
```diff
@@ -692,6 +694,28 @@ def norm_encoder_hidden_states(self, encoder_hidden_states: torch.Tensor) -> tor
 
         return encoder_hidden_states
 
+    @torch.no_grad()
+    def enable_fused_projections(self, device, dtype, fuse_projections=True, is_cross_attention=False):
```
Comment (patrickvonplaten) [inline src/diffusers/models/attention_processor.py:698]:
```suggestion
    def enable_fused_projections(self, device, dtype, fuse_projections=True):
```

Can't we retrieve whether we're in cross attention mode just from: `self.cross_attention_dim == self.query_dim` (we probably need to save query_dim in init)

Also why do we need to pass device and dtype? Can't we just retrieve it from the prev projection layers?

File: src/diffusers/models/attention_processor.py [hunk 6]
```diff
@@ -1253,6 +1274,101 @@ def __call__(
...
+    <Tip warning={true}>
+
+    This API is currently 🧪 experimental in nature and can change in future.
+
+    </Tip>
+    """
+
+    def __init__(self):
+        if not hasattr(F, "scaled_dot_product_attention"):
+            raise ImportError("FusedAttnProcessor2_0 requires PyTorch 2.0, to use it, please upgrade PyTorch to 2.0.")
```
Comment (patrickvonplaten) [inline src/diffusers/models/attention_processor.py:1292]:
```suggestion
            raise ImportError("FusedAttnProcessor2_0 requires at least PyTorch 2.0, to use it. Please upgrade PyTorch to > 2.0.")
```

File: src/diffusers/schedulers/scheduling_euler_discrete.py [hunk 7]
```diff
@@ -289,6 +290,8 @@ def set_timesteps(self, num_inference_steps: int, device: Union[str, torch.devic
             self.timesteps = torch.from_numpy(timesteps.astype(np.float32)).to(device=device)
 
         self.sigmas = torch.cat([sigmas, torch.zeros(1, device=sigmas.device)])
+        if "cuda" in str(sigmas.device):
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/schedulers/scheduling_euler_discrete.py:293]: It helps to have a nice compiled graph. It contributes to the speedup.

**Comment by sayakpaul** [comment]
> Looks very nice! Could you add some code that would allow to test the speed-ups?

https://github.com/sayakpaul/sdxl-fast

**Comment by patrickvonplaten** [comment]
Indeed, getting a nice 5% speed-up!

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py [hunk 8]
```diff
@@ -778,6 +838,26 @@ def denoising_end(self):
...
+    #         for name, module in model.named_children():
+    #             if isinstance(module, torch.nn.GroupNorm):
+    #                 new_gn = GroupNorm32(module.num_groups, module.num_channels, module.eps, module.affine)
+    #                 setattr(model, name, new_gn)
+    #             elif len(list(module.children())) > 0:
+    #                 recursive_fn(module)
+
+    #     recursive_fn(self.unet)
+    #     recursive_fn(self.vae)
+
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:860]:
```suggestion
```

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py [hunk 9]
```diff
@@ -1187,6 +1267,8 @@ def __call__(
                 self.upcast_vae()
                 latents = latents.to(next(iter(self.vae.post_quant_conv.parameters())).dtype)
 
+            # if hasattr(self, "is_vae_in_blfoat16") and self.is_vae_in_blfoat16:
+            #     latents = latents.to(self.vae.dtype)
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:1271]:
```suggestion
```

**Review by patrickvonplaten — APPROVED** [review]
Cool feature!

**Comment by sayakpaul** [comment]
@DN6 any reason why the fetcher is failing?

**Review by patrickvonplaten — APPROVED** [review]
Very nice job here!

File: src/diffusers/schedulers/scheduling_euler_discrete.py [hunk 10]
```diff
@@ -289,6 +290,8 @@ def set_timesteps(self, num_inference_steps: int, device: Union[str, torch.devic
             self.timesteps = torch.from_numpy(timesteps.astype(np.float32)).to(device=device)
 
         self.sigmas = torch.cat([sigmas, torch.zeros(1, device=sigmas.device)])
+        if sigmas.device.type == "cuda":
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_euler_discrete.py:293]: this breaks the `add_noise` function and thus inpaint and img2img and training

**Pull request merged**